src/msg/producer/writer/shard_writer.go

```diff
@@ -160,6 +160,12 @@ func newReplicatedShardWriter(
 
 func (w *replicatedShardWriter) Write(rm *producer.RefCountedMessage) {
 	w.RLock()
+	if len(w.messageWriters) == 0 {
+		w.RUnlock()
+		w.logger.Error("no message writers available for shard", zap.Int("shard", int(rm.Shard())))
+		w.m.noWritersError.Inc(1)
```

Comment:
Nit: might be slightly better to increase the metrics before logging, we have seen sync logging getting blocked sometimes.

Reply:
https://github.com/m3db/m3/pull/3558/commits/808c4b708620664e7a8bc5a9d753aa1df7864b9a